src/main/java/org/opensearch/geospatial/ip2geo/action/UpdateDatasourceTransportAction.java

```diff
@@ -82,16 +86,23 @@ protected void doExecute(final Task task, final UpdateDatasourceRequest request,
                 );
                 return;
             }
-
             try {
-                Datasource datasource = datasourceFacade.getDatasource(request.getName());
-                if (datasource == null) {
-                    throw new ResourceNotFoundException("no such datasource exist");
-                }
-                validate(request, datasource);
-                updateIfChanged(request, datasource);
-                lockService.releaseLock(lock);
-                listener.onResponse(new AcknowledgedResponse(true));
+                // TODO: makes every sub-methods as async call to avoid using a thread in generic pool
+                threadPool.generic().submit(() -> {
+                    try {
+                        Datasource datasource = datasourceFacade.getDatasource(request.getName());
+                        if (datasource == null) {
```

Comment:
if this check costly? If check is failed we can skip the overhead of spinning new thread worker? I think it's better f we can check this before submitting a new job.

Reply:
Not costly but I will work on it as a follow up action item. It is commented as TODO.